src/operators.jl

```diff
@@ -595,6 +595,217 @@ function tensor_product(g::G, h::G) where {G<:AbstractGraph}
     return z
 end
 
+"""
+    strong_product(g, h)
+
+Return the [strong product](https://en.wikipedia.org/wiki/Strong_product_of_graphs)
+of `g` and `h`.
+
+### Implementation Notes
+Preserves the eltype of the input graph. Will error if the number of vertices
+in the generated graph exceeds the eltype.
+
+# Examples
+```jldoctest
+julia> using Graphs
+
+julia> a = star_graph(3)
+
+julia> b = path_graph(3)
+
+julia> g = strong_product(a, b)
+{9, 8} undirected simple Int64 graph
+
+julia> g == union(cartesian_product(a, b), tensor_product(a, b))
+true
+```
+"""
+function strong_product(g::G, h::G) where {G<:AbstractGraph}
```

Comment:
Does this operator also work for directed graphs? And does it really make sense to define it for all graph types? Maybe we should restrict it to `SimpleGraph` and maybe also `SimpleDiGraph`.

Otherwise it might be a bit confusing if one has, for example, a graph with metadata on its vertices.

Reply:
Though I don't see mention in wikipedia of directed graphs in relation to any of these products, I think most of them would make sense.  The only exception would be the homomorphic product which has (h_1 \nsim h_2) in its definition and there could be controversy over how that should be interpreted for directed graphs.  Also this one is pretty obscure and I suspect nobody would ever want to use it on a directed graph.

Good point on not supporting graphs that have metadata.  What would be the appropriate way to support both SimpleGraph and SimpleDiGraph?  Should I use the common parent class AbstractSimpleGraph?  (Note that the existing functions tensor_product and cartesian_product also have this issue.)